[VAULTS] Accounting Oracle

tamtamchik opened the pull request:
Base PR for Lido V3 accounting oracle

F4ever commented on src/modules/accounting/types.py:
```diff
@@ -132,18 +133,68 @@ class WithdrawalRequestStatus:
 type RebaseReport = tuple[ValidatorsCount, ValidatorsBalance, WithdrawalVaultBalance, ELVaultBalance, SharesToBurn]
-type WqReport = tuple[BunkerMode, FinalizationShareRate, FinalizationBatches]
+type WqReport = tuple[BunkerMode, FinalizationBatches]
+type VaultsReport = tuple[list[int], list[int]]
```
Create custom type like

F4ever commented on src/providers/execution/contracts/accounting.py:
```diff
@@ -0,0 +1,168 @@
+class AccountingContract(ContractInterface):
+    abi_path = './assets/Accounting.json'
+
+    def handle_oracle_report(
```
simplify this func

F4ever commented on src/providers/execution/contracts/vault_hub.py:
```diff
@@ -0,0 +1,46 @@
+            'to': self.address,
+        })
+
+        return response
```
add `get_all_vaults` method

tamtamchik replied:
Done in 49fdc6c7

F4ever commented on src/web3py/extensions/staking_vaults.py:
```diff
@@ -0,0 +1,149 @@
+import logging
```
Move some code to Vault contract and some to separate service

sergeyWh1te replied:
done

F4ever commented on src/providers/consensus/client.py:
```diff
@@ -174,6 +175,12 @@ def get_validators(self, blockstamp: BlockStamp) -> list[Validator]:
+    def get_pending_deposits(self, blockstamp: BlockStamp) -> list[PendingDeposit]:
+        return self.get_state_view(blockstamp).pending_deposits
+
+    def get_pending_deposits_no_cache(self, blockstamp: BlockStamp) -> list[PendingDeposit]:
```
Remove this

sergeyWh1te replied:
done

sergeyWh1te commented:
1) collapse: abi
2) transfer staking_vault -> modules/accounting
3) contracts loading moved -> web3py/extentions/contract
4) AccountingContract.handle_oracle_report - place as is - agreed with @F4ever that they'll fix that
5) Integration tests - are broken. It's not trivial work - agreed with @F4ever that they or we'll collaborate how to fix tests when @F4ever will figure out "how" to fix them

cc @tamtamchik @vp4242

sergeyWh1te commented:
<img width="1466" alt="Screenshot 2025-04-09 at 3 43 07 PM" src="https://github.com/user-attachments/assets/26d84bba-d500-4699-b9d0-7206ac063fcb" />

On local machine with sepolia urls

chasingrainbows commented on src/main.py:
```diff
@@ -75,20 +79,24 @@ def main(module_name: OracleModule):
     logger.info({'msg': 'Check configured providers.'})
-    if Version(kac.get_status().appVersion) < constants.ALLOWED_KAPI_VERSION:
-        raise IncompatibleException(f'Incompatible KAPI version. Required >= {constants.ALLOWED_KAPI_VERSION}.')
+    if Version(kac.get_status().appVersion) < Version('1.5.0'):
```
i guess there is broken merge

```
    if Version(kac.get_status().appVersion) < constants.ALLOWED_KAPI_VERSION:
        raise IncompatibleException(f'Incompatible KAPI version. Required >= {constants.ALLOWED_KAPI_VERSION}.')
```
is more actual

sergeyWh1te replied:
Merged from oracleV6

chasingrainbows commented on src/main.py:
```diff
@@ -35,14 +35,16 @@
-        },
-    })
+    logger.info(
+        {
```
There are too many formatting changes that don't make sense in terms of business logic. If you are automatically using black/another formatter, please don't, as it makes the review process more difficult. For now black is used only for tests by the oracle team.

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,244 @@
+        deposit_map = dict[str, int]()
+
+        for deposit in pending_deposits:
+            if deposit.withdrawal_credentials not in deposit_map:
```
U can use defaultdict here 

```
    deposit_map: dict[str, int] = defaultdict(int)
    for deposit in pending_deposits:
        deposit_map[deposit.withdrawal_credentials] += Web3.to_wei(int(deposit.amount), 'gwei')
```

sergeyWh1te replied:
done

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,244 @@
+
+            vault = self._load_vault(vault_socket.vault)
+            vault_in_out_delta = vault.in_out_delta(blockstamp)
+            print(vault_in_out_delta)
```
Plz remove prints or replace with logging

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,244 @@
+
+            pending_deposit = 0
+            if vault_withdrawal_credentials in deposit_map:
+                pending_deposit = deposit_map[vault_withdrawal_credentials]
```
Hmmm, looks like a bug, is there 
```suggestion
                pending_deposit += deposit_map[vault_withdrawal_credentials]
```
supposed to be?

chasingrainbows replied:
If yes, there is test for this one need to add

sergeyWh1te replied:
changed

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,244 @@
+
+    @staticmethod
+    def connect_vault_to_validators(validators: list[Validator], vault_addresses: VaultsMap) -> VaultToValidators:
+        wc_vault_map = dict[str, VaultData]()
```
```suggestion
    wc_vault_map: dict[str, VaultData] = {
        vault_data.withdrawal_credentials: vault_data
        for vault_data in vault_addresses.values()
    }

```

sergeyWh1te replied:
fixed

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,244 @@
+        for vault_pk in vault_addresses:
+            wc_vault_map[vault_addresses[vault_pk].withdrawal_credentials] = vault_addresses[vault_pk]
+
+        result = VaultToValidators()
```
defaultdict will make a little bit cleaner here

```suggestion
    result: VaultToValidators = defaultdict(list)
    for validator in validators:
        wc = validator.validator.withdrawal_credentials
        if wc in wc_vault_map:
            result[wc_vault_map[wc].address].append(validator)
```

sergeyWh1te replied:
fixed

chasingrainbows commented on src/modules/accounting/types.py:
```diff
@@ -28,8 +28,9 @@ class ReportData:
     withdrawal_finalization_batches: list[int]
-    finalization_share_rate: int
     is_bunker: bool
+    vaults_values: list[int]  # -> TODO replace to TreeHash
```
TODO? Why not impement this now?

chasingrainbows commented on tests/conftest.py:
```diff
@@ -117,8 +118,8 @@ def csm(web3):
 def contracts(web3, provider):
-    src.variables.LIDO_LOCATOR_ADDRESS = "0x548C1ED5C83Bdf19e567F4cd7Dd9AC4097088589"
-    src.variables.CSM_MODULE_ADDRESS = "0xdA7dE2ECdDfccC6c3AF10108Db212ACBBf9EA83F"
+    src.variables.LIDO_LOCATOR_ADDRESS = variables.LIDO_LOCATOR_ADDRESS
```
This change is senseless, src.variables.LIDO_LOCATOR_ADDRESS == variables.LIDO_LOCATOR_ADDRESS, result will be the same if we just remove this line.

You need something like
```suggestion
if not src.variables.LIDO_LOCATOR_ADDRESS:
    src.variables.LIDO_LOCATOR_ADDRESS = "0x548C1ED5C83Bdf19e567F4cd7Dd9AC4097088589"
```
Otherwise this will break our tests on mainnet after merge

sergeyWh1te replied:
fixed

chasingrainbows commented on tests/modules/accounting/staking_vault/test_staking_vaults.py:
```diff
@@ -0,0 +1,145 @@
+        expected = '0x8947e4ef0354707240394a85ca9c2dcad15f52c773a955a0b827f50f5afbb93b'
+        assert got == expected
+
+        # proof_cid = self.StakingVaults.publish_proofs(merkle_tree, bs, vaults)
```
I guess comments not actual anymore

sergeyWh1te commented on src/web3py/extensions/staking_vaults.py:
```diff
@@ -0,0 +1,149 @@
+            ),
+        )
+
+        self.vault_hub: VaultHubContract = cast(
```
moved

chasingrainbows commented on src/main.py:
```diff
@@ -90,6 +91,8 @@ def main(module_name: OracleModule):
         'ipfs': lambda: ipfs,  # type: ignore[dict-item]
     })
 
+    web3.staking_vaults = StakingVaults(web3, cc, ipfs, web3.lido_contracts.vault_hub)
```
Better to do it though attach_modules

chasingrainbows replied:
Right way to do this - though web3.attach_modules. You dont need to pass cc/ipfs/web3.lido_contracts.vault_hub separately, just provide web3 object from here https://github.com/lidofinance/lido-oracle/blob/d81b3f3deb0ff281b5433b534add9d8df5ef1900/src/web3py/types.py#L16-L16, then use cc/ipfs/etc inside StakingVaults though web3 object. 
```
# example
self.w3.cc.<method u need>()
```
Here is example from codebase: https://github.com/lidofinance/lido-oracle/blob/d81b3f3deb0ff281b5433b534add9d8df5ef1900/src/web3py/extensions/lido_validators.py#L125-L125

chasingrainbows commented on src/modules/accounting/accounting.py:
```diff
@@ -397,7 +448,11 @@ def _combine_report_parts(
             shares_requested_to_burn=shares_requested_to_burn,
             withdrawal_finalization_batches=finalization_batches,
-            finalization_share_rate=finalization_share_rate,
+            # TODO: fees and deficit are not implemented yet
```
If we cant remove this "todo" now, better to leave some task id or pr link in comment maybe, to fix this in future

chasingrainbows commented:
The overall look is good! I believe merging with https://github.com/lidofinance/lido-oracle/pull/698 til the end of next week once it's completed will help the tests pass.

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,372 @@
+
+        return tree_data, vaults
+
+    def get_vaults(self, blockstamp: BlockStamp) -> VaultsMap:
```
could be private `_get_vaults`

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,372 @@
+        return out
+
+    @staticmethod
+    def connect_vaults_to_validators(validators: list[Validator], vault_addresses: VaultsMap) -> VaultToValidators:
```
could be private

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,372 @@
+        return result
+
+    @staticmethod
+    def connect_vaults_to_pending_deposits(
```
could be private

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,372 @@
+        return StandardMerkleTree(data, ("address", "uint256", "uint256", "uint256", "uint256"))
+
+    @staticmethod
+    def get_vault_to_proof_map(merkle_tree: StandardMerkleTree, vaults: VaultsMap) -> dict[str, VaultProof]:
```
could be private

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,372 @@
+
+        return result
+
+    def publish_proofs(self, tree: StandardMerkleTree, bs: BlockStamp, vaults: VaultsMap) -> CID:
```
could be private

chasingrainbows commented on tests/modules/accounting/test_accounting_module.py:
```diff
@@ -440,13 +462,96 @@ def test_simulate_rebase_after_report(
-    accounting.w3.lido_contracts.lido.handle_oracle_report = Mock(return_value=LidoReportRebaseFactory.build())  # type: ignore
+    accounting.w3.lido_contracts.accounting = Mock()
+
+    accounting.w3.lido_contracts.accounting.simulate_oracle_report = Mock(
```
If u mocked some method, for example here, u have to call then `accounting.w3.lido_contracts.accounting.simulate_oracle_report.assert_called_with(<expected arguments>)`, because there is no guarantees that mocked method was called with something that u expect while it is mocked.

chasingrainbows commented on src/providers/execution/contracts/vault_hub.py:
```diff
@@ -0,0 +1,125 @@
+        return response
+
+    @lru_cache(maxsize=1)
+    def vault_socket(self, vault_id: int, block_identifier: BlockIdentifier = 'latest') -> VaultSocket:
```
Any tests for this one?

sergeyWh1te replied:
added test

chasingrainbows commented on src/providers/execution/contracts/staking_vault.py:
```diff
@@ -0,0 +1,50 @@
+logger = logging.getLogger(__name__)
+
+
+class StakingVaultContract(ContractInterface):
```
Cant found any usage for methods of this contract, is it still actual?

tamtamchik replied:
Fixed in f2ddaf2

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,372 @@
+        self.cl = cl
+        self.vault_hub = vault_hub
+
+    def _load_vault(self, address: ChecksumAddress) -> StakingVaultContract:
```
Cant find any calls

tamtamchik replied:
Fixed in f2ddaf2

sergeyWh1te commented on src/providers/execution/contracts/vault_hub.py:
```diff
@@ -0,0 +1,125 @@
+            response.pendingDisconnect,
+        )
+
+    def get_report(self, block_identifier: BlockIdentifier = 'latest'):
```
added test

sergeyWh1te committed:
fix(vauts): restore unit tests

sergeyWh1te committed:
fix(vaults): run black for tests

sergeyWh1te committed:
fix(vaults): tests

sergeyWh1te committed:
fix(vaults): tests

sergeyWh1te committed:
fix(vaults): fix pylint

tamtamchik committed:
feat: recalculate fees

sergeyWh1te committed:
tets(vaults): trying get apr

tamtamchik committed:
Merge branch 'feat/vaults-pectra' of github.com:lidofinance/lido-oracle into feat/vaults-pectra

# Conflicts:
#	src/modules/accounting/accounting.py

sergeyWh1te committed:
tets(vaults): trying get apr

sergeyWh1te committed:
tets(vaults): trying get apr

sergeyWh1te committed:
tets(vaults): trying get apr

tamtamchik committed:
chore: rearrange fees and take vault_liquidity_fee into account

tamtamchik committed:
chore: update fees calculation

sergeyWh1te committed:
tets(vaults): get fees

sergeyWh1te committed:
Merge branch 'feat/vaults-pectra' of work:lidofinance/lido-oracle into feat/vaults-pectra

sergeyWh1te committed:
feat(vaults): fees review

sergeyWh1te committed:
feat(vaults): removed publish proof

sergeyWh1te committed:
feat(vaults): removed parsing proofs

sergeyWh1te committed:
fix(vaults): parsing extraValue

sergeyWh1te committed:
fix(vaults): write inOutDelta as str

sergeyWh1te committed:
fix(vaults): format leafIndexToData

sergeyWh1te committed:
test(vaults): fee

sergeyWh1te committed:
test(vaults): set reserve_ratioBP, forced_rebalance_thresholdBP, infra_feeBP, liquidity_feeBP, reservation_feeBP

sergeyWh1te committed:
test(vaults): set liability_shares for prev report

tamtamchik committed:
chore: after fees review

tamtamchik committed:
chore: fix mapping error

sergeyWh1te committed:
test(vaults): fix actual fees for staking vaults

sergeyWh1te committed:
test(vaults): skip TestStakingVaultsSmoke

sergeyWh1te committed:
test(vaults): skip TestStakingVaultsSmoke

sergeyWh1te committed:
test(vaults): fix test_hash_calculations

chasingrainbows committed:
feat(vault-pectra): move part of test_lido_locator_contract to testnet

chasingrainbows committed:
Merge branch 'feat/vaults-pectra' of github.com:lidofinance/lido-oracle into feat/vaults-pectra

sergeyWh1te committed:
refactor(vaults): Add test_infra_fee, test_reservation_liquidity_fee

sergeyWh1te committed:
fix(vaults): calculation of liquidity_fee. Missed last sector for calculation

sergeyWh1te committed:
feat(vaults): calc fees

TheDZhon commented on src/modules/accounting/accounting.py:
```diff
@@ -242,60 +267,66 @@ def _get_finalization_data(self, blockstamp: ReferenceBlockStamp) -> tuple[Final
 
-        return self.w3.lido_contracts.lido.handle_oracle_report(
-            # Lido contract has sanity check that timestamp is not in the future.
+        withdrawal_share_rate = 0  # For initial calculation we assume 0 share rate
```
```suggestion
        withdrawal_share_rate = 0  # For simulation we assume 0 share rate
```

TheDZhon commented on src/modules/accounting/accounting.py:
```diff
@@ -393,9 +460,176 @@ def _combine_report_parts(
+
+    # pylint: disable=too-many-branches,too-many-statements
+    def _get_vaults_fees(self, blockstamp: ReferenceBlockStamp, vaults: VaultsMap, vaults_total_values: list[int]) -> list[int]:
+        prev_report = self.w3.staking_vaults.get_prev_report(blockstamp)
```
```suggestion
        prev_report = self.w3.staking_vaults.get_prev_ipfs_report(blockstamp)
```

TheDZhon commented on src/modules/accounting/accounting.py:
```diff
@@ -393,9 +460,176 @@ def _combine_report_parts(
+        prev_report = self.w3.staking_vaults.get_prev_report(blockstamp)
+        ## When we do not have a report - then we do not have starting point for calculation
+        if prev_report is None:
+            rebased_event = self.w3.lido_contracts.lido.get_last_token_rebased_event(blockstamp.block_number - 7200, blockstamp.block_number)
```
we shall not rely on magic 7200 number and rather retrieve the latest delivered report

https://docs.lido.fi/contracts/accounting-oracle#getlastprocessingrefslot

TheDZhon commented on src/modules/accounting/accounting.py:
```diff
@@ -393,9 +460,176 @@ def _combine_report_parts(
+        slots_elapsed = self._get_slots_elapsed_from_last_report(blockstamp)
+        time_elapsed = slots_elapsed * chain_conf.seconds_per_slot
+
+        core_apr_ratio: float = 0.0
```
let's move to ray math

TheDZhon replied:
https://github.com/dapphub/ds-math/blob/master/src/math.sol

TheDZhon commented on src/modules/accounting/accounting.py:
```diff
@@ -393,9 +460,176 @@ def _combine_report_parts(
+
+        events = defaultdict(list)
+        fees_updated_events_map = defaultdict(list)
+        fees_updated_events = self.w3.lido_contracts.vault_hub.get_vaults_fee_updated_events(prev_block_number,                                                             blockstamp.block_number)
```
looooooooooooong

TheDZhon reviewed (commented):
:eyes:

sergeyWh1te committed:
feat(vaults): fix review attempt 1

sergeyWh1te committed:
feat(vaults): temp commit for debugging

sergeyWh1te committed:
feat(vaults): cast prevFee to int

sergeyWh1te committed:
feat(vaults): Add decimal for calc fees

sergeyWh1te committed:
test(vaults): cross testing for taking fees

sergeyWh1te committed:
test(vaults): cross testing for taking fees

sergeyWh1te committed:
feat(vaults): get previous report from last processing refSlot

sergeyWh1te committed:
fix(vaults): calc vault fees on mainet, tesnet at the first time

sergeyWh1te committed:
fix(vaults): calc vault fees on mainet, tesnet at the first time

sergeyWh1te committed:
fix(vaults): calc vault fees on mainet, tesnet at the first time

sergeyWh1te committed:
fix(vaults): rename constants for getting slashing period

sergeyWh1te committed:
fix(vaults): use PRECISION_E27 for taking fees

sergeyWh1te committed:
feat(vaults): move fees into staking_vaults

sergeyWh1te committed:
chore(vaults): black formatting

sergeyWh1te committed:
chore(vaults): black formatting

sergeyWh1te committed:
fix(vaults): test apr by using 27 decimals instead of 28

TheDZhon reviewed (commented):
:eyes:

sergeyWh1te committed:
fix(vaults): review marks

sergeyWh1te committed:
fix(vaults): review marks

sergeyWh1te committed:
fix(vaults): review marks

sergeyWh1te committed:
chore(vaults): fix lint

sergeyWh1te committed:
chore(vaults): fix lint

sergeyWh1te committed:
fix(vaults): fix VaultInfo getting inOutDelta

sergeyWh1te committed:
chore(vaults): lint

sergeyWh1te committed:
chore(vaults): lint

TheDZhon commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,635 @@
+
+        dumped_tree_str = json.dumps(output, default=encoder)
+
+        cid = self.ipfs_client.publish(dumped_tree_str.encode('utf-8'), 'merkle_tree.json')
```
probably can consider adding backups to GH or somewhere

TheDZhon commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,635 @@
+            ),
+        )
+        initial_ref_slot = hash_consensus.get_initial_ref_slot(blockstamp.block_hash)
+        # If skipped, we reference the block from the first non-missed slot (+100 offset guarantees availability).
```
here we expect that genesis devnet might have missed slots which is acknowledged

TheDZhon commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,635 @@
+        vault_pending_deposits: list[PendingDeposit],
+        vault_withdrawal_credentials: str,
+    ) -> int:
+        validator_pubkeys = set(validator.validator.pubkey for validator in validators)
```
computationally heavy

TheDZhon commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,635 @@
+
+            # Case 2: Validator exists but not bound to this vault
+            if pubkey in validator_pubkeys:
+                validator = next(v for v in validators if v.validator.pubkey == pubkey)
```
:thinking:

TheDZhon commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,635 @@
+            # Case 2: Validator exists but not bound to this vault
+            if pubkey in validator_pubkeys:
+                validator = next(v for v in validators if v.validator.pubkey == pubkey)
+                if validator.validator.withdrawal_credentials == vault_withdrawal_credentials:
+                    total_value += deposit_value
```
dead code?

TheDZhon commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,635 @@
+                minted_steth, blocks_elapsed, core_apr_ratio, liquidity_fee
+            )
+        elif len(events[vault_address]) > 0:
+            # In case of events, we iterate through them backwards, calculating liquidity fee for each interval based
```
can we add ASCII art

TheDZhon reviewed (changes requested):
:+1: Great effort so far! Probably can finalize it quite soonish :pray:

sergeyWh1te committed:
feat(vaults): add split up fees into ipfs report

sergeyWh1te committed:
chore(vaults): lint

sergeyWh1te committed:
feat(vaults): removed merkle_tree_root from ipfs_report

sergeyWh1te committed:
fix(vaults): get starting period for previous report

sergeyWh1te committed:
test(vaults): add chain, frame_config

sergeyWh1te committed:
test(vaults): add chain, frame_config

sergeyWh1te committed:
fix(vaults): calc pending deposits

sergeyWh1te committed:
fix(vaults): log comments

sergeyWh1te committed:
feat(vaults): Add ASCII for calculation liquidity_fee

sergeyWh1te committed:
feat(vaults): Add ASCII for calculation liquidity_fee

sergeyWh1te committed:
chore(vaults): lint

sergeyWh1te committed:
feat(vaults): refactoring events

TheDZhon commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,669 @@
+                    continue
+
+                if isinstance(event, BadDebtSocializedEvent):
+                    if vault_address == event.vault_donor:
```
:thinking:

TheDZhon commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,669 @@
+                    liability_shares += event.amount_of_shares
+                    current_block = event.block_number
+                    continue
+                if isinstance(event, MintedSharesOnVaultEvent):
+                    liability_shares -= event.amount_of_shares
+                    current_block = event.block_number
```
let's move this upper (to be the first case)

TheDZhon commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,669 @@
+            if pubkey in vault_validator_pubkeys:
+                total_value += deposit_value
+                continue
+
```
:warning: :warning: we should still have `continue` case when the pubkey belong to the non-vaults validator set

TheDZhon commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,669 @@
+
+        ## Fresh devnet
+        ## We DO not have prev IPFS report, and we DO not have prev Oracle report then we take
+        # If skipped, we reference the block from the first non-missed slot (frame length offset guarantees availability).
```
```suggestion
        # If skipped, we reference the block from the first non-missed slot (frame length offset presumes availability).
```

TheDZhon reviewed (changes requested):
:+1:

sergeyWh1te committed:
feat(vaults): fix review  get_vaults_total_values

chasingrainbows commented on tests/conftest.py:
```diff
@@ -100,8 +101,11 @@ def configure_testnet_tests(request, monkeypatch):
+        # TODO: revert to https://docs.lido.fi/deployed-contracts/hoodi after vaults deploy
+        # monkeypatch.setattr(variables, 'LIDO_LOCATOR_ADDRESS', LIDO_LOCATOR_ADDRESS)
+        monkeypatch.setattr(variables, 'LIDO_LOCATOR_ADDRESS', '0xD7c1B80fA86965B48cCA3aDcCB08E1DAEa291980')
+        monkeypatch.setattr(variables, 'CSM_MODULE_ADDRESS', CSM_MODULE_ADDRESS)
```
Hardcoded testnet address must be used.

chasingrainbows commented on pyproject.toml:
```diff
@@ -28,6 +28,9 @@ web3-multi-provider = { version = "^2.2.1", extras = ["metrics"] }
 py-multiformats-cid = "^0.4.4"
+ssz = "^0.5.2"
+py-ecc = "^8.0.0"
+py-evm = "^0.12.1b1"
```
Tests can drop some warnings without this dependency, but we dont use it and dont need it.

tamtamchik replied:
it looks like it is not used anymore, and should be removed. 
cc @chasingrainbows @sergeyWh1te

sergeyWh1te committed:
feat(vaults): fix review marks

sergeyWh1te committed:
feat(vaults): fix review marks

sergeyWh1te committed:
feat(vaults): fix review marks

sergeyWh1te committed:
feat(vaults): try use hardcoded CSM_MODULE_ADDRESS

chasingrainbows commented on poetry.lock:
```diff
@@ -1,12 +1,12 @@
-# This file is automatically @generated by Poetry and should not be changed by hand.
+# This file is automatically @generated by Poetry 2.1.1 and should not be changed by hand.
```
Current production version pinned at docker - 1.3.2. You can regenerate poetry.lock version-aware using command `make poetry-lock`

sergeyWh1te committed:
feat(vaults): try use hardcoded CSM_MODULE_ADDRESS

sergeyWh1te committed:
feat(vaults):  make poetry-lock

F4ever commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,676 @@
+        super().__init__(w3)
+
+        self.w3 = w3
+        self.ipfs_client = ipfs
+        self.cl = cl
+        self.lido = lido
+        self.vault_hub = vault_hub
+        self.lazy_oracle = lazy_oracle
+        self.accounting_oracle = accounting_oracle
```
Let's remove this code.
You can access to this instances via 
```
self.w3.ipfs
self.w3.cl
self.w3.lido_contracts.lido
...
```

F4ever commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,676 @@
+from web3.types import Wei
+
+from src.constants import TOTAL_BASIS_POINTS, PRECISION_E27, EPOCHS_PER_SLASHINGS_VECTOR
+from src.modules.accounting.events import VaultFeesUpdatedEvent, BurnedSharesOnVaultEvent, MintedSharesOnVaultEvent, \
```
Pls, use brackets instead of \

sergeyWh1te commented on poetry.lock:
```diff
@@ -1,12 +1,12 @@
-# This file is automatically @generated by Poetry and should not be changed by hand.
+# This file is automatically @generated by Poetry 2.1.1 and should not be changed by hand.
```
already done

sergeyWh1te commented on src/main.py:
```diff
@@ -89,6 +90,14 @@ def main(module_name: OracleModule):
         'ipfs': lambda: ipfs,  # type: ignore[dict-item]
     })
 
+    web3.staking_vaults = StakingVaults(
```
Cause no visible approach how to submodule for accounting - I'll jump to dm how to troubleshoot this.

chasingrainbows commented on src/modules/accounting/accounting.py:
```diff
@@ -393,9 +495,46 @@ def _combine_report_parts(
+        post_total_pooled_ether: int,
+    ) -> Decimal:
+        with localcontext() as ctx:
+            ctx.prec = PRECISION_E27
```
What about rounding? Looks like without explicit rounding, the protocol may undercharge fees due to precision loss in decimal division. Python's Decimal default rounding mode is ROUND_HALF_EVEN, should we use ROUND_UP?

sergeyWh1te replied:
yes

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,676 @@
+                return asdict(o)
+            raise TypeError(f"Object of type {type(o)} is not JSON serializable")
+
+        def stringify_values(data) -> list[dict[str, Any]]:
```
Looks liike Any could be replaced with more strict annotation here

sergeyWh1te committed:
feat(vaults): fix review

sergeyWh1te committed:
feat(vaults): fix review

sergeyWh1te committed:
feat(vaults): fix review

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,660 @@
+        liability_shares (Y)
+                ↑
+                │
+                │
```
Stricter typing needs here

sergeyWh1te replied:
?

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,660 @@
+                if isinstance(event, BadDebtWrittenOffToBeInternalizedEvent):
+                    liability_shares += event.bad_debt_shares
+                    current_block = event.block_number
+                    continue
```
Is continue missed here? Any test for this case?

sergeyWh1te replied:
Missed. fixed. thanks

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,660 @@
+                    continue
+
+                if isinstance(event, BadDebtWrittenOffToBeInternalizedEvent):
+                    liability_shares += event.bad_debt_shares
```
elif instead of each next if shoud be used

sergeyWh1te committed:
feat(vaults): fix review

sergeyWh1te committed:
feat(vaults): fix review

sergeyWh1te committed:
feat(vaults): fix review

sergeyWh1te committed:
feat(vaults): fix typing

sergeyWh1te committed:
feat(vaults): fix review

sergeyWh1te committed:
test(vaults): add get_staking_fee_aggregate_distribution

TheDZhon reviewed (commented):
:star2: Let's try to deploy the oracle to testnet-2 on Friday hopefully sending the frozen commit to auditors on Monday :pray:

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,655 @@
+        vault_address: str,
+        liability_shares: Shares,
+        liquidity_fee_bp: int,
+        events: dict,
```
Stricter typing needs here

chasingrainbows commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,655 @@
+            )
+
+            prev_liability_shares = 0
+            if vaults_on_prev_report.get(vault_address) is not None:
```
There is no need to suppress linter
```
vault = vaults_on_prev_report.get(vault_address)
if vault is not None:
    prev_liability_shares = vault.liability_shares
```
or
```
if vault := vaults_on_prev_report.get(vault_address):
    prev_liability_shares = vault.liability_shares
```

chasingrainbows reviewed (changes requested):
Thx, great job! Left some threads behind.

vp4242 committed:
feat: slashing_reserve, ipfs_report verification, genesis fork

vp4242 committed:
test: fix tests for mainnet

vp4242 committed:
test: extract common consts

F4ever commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,671 @@
+        for vault_address, vault in vaults.items():
+            out[vault_address] = vault.balance
+            vault_validators = vaults_validators.get(vault_address, [])
+            vault_pending_deposits = vaults_pending_deposits.get(vault_address, [])
```
Let's get value by key. If key does not exist let's Oracle fail with error.

TheDZhon commented on pyproject.toml:
```diff
@@ -28,6 +28,9 @@ web3-multi-provider = { version = "^2.2.1", extras = ["metrics"] }
 oz-merkle-tree = { git = "https://github.com/lidofinance/oz-merkle-tree" }
 py-multiformats-cid = "^0.4.4"
+ssz = "^0.5.2"
+py-ecc = "^8.0.0"
```
We need this lib, let's pin it appropriately and highlight in the audit scope that it's appeared

sergeyWh1te committed:
fix(vaults): removed cache under get_vaults_slashing_reserve

sergeyWh1te committed:
feat(vaults): restore calc pending deposits

sergeyWh1te committed:
fix(vaults): removed annotation under getting vaults

sergeyWh1te commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,676 @@
+
+        return cid
+
+    def get_ipfs_report(self, ipfs_report_cid: str) -> MerkleTreeData:
```
We grab data from onchain. Then check everything down

sergeyWh1te commented on src/modules/accounting/types.py:
```diff
@@ -130,20 +141,167 @@ class WithdrawalRequestStatus:
+
+@dataclass
+class MerkleValue:
+    vault_address: str
+    total_value_wei: int
```
vault_address its'a string after parsing

sergeyWh1te commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,676 @@
+            vault_data.withdrawal_credentials: vault_data for vault_data in vault_addresses.values()
+        }
+
+        result: VaultToValidators = defaultdict(list)
```
It break typing with VaultToValidators

sergeyWh1te commented on src/modules/accounting/staking_vaults.py:
```diff
@@ -0,0 +1,671 @@
+                continue
+
+            # Case 3: No validator found for this pubkey - validate deposits
+            deposits_for_pubkey = [d for d in pending_deposits if d.pubkey == pubkey]
```
Wait @tamtamchik pls

TheDZhon replied:
should be named as:
- deposits_for_pubkey → all_deposits_for_pubkey
- deposits_by_pubkey → vault_deposits_by_pubkey

sergeyWh1te committed:
feat(vaults): fix review

sergeyWh1te committed:
feat(vaults): fix review

sergeyWh1te committed:
feat(vaults): fix review

sergeyWh1te committed:
test(vaults): black formatter

sergeyWh1te committed:
fix(vaults): tests

sergeyWh1te committed:
feat(vaults): review fixes

sergeyWh1te committed:
feat(vaults): review fixes

sergeyWh1te committed:
feat(vaults): review fixes

sergeyWh1te committed:
feat(vaults): fix getting previous report start point

sergeyWh1te committed:
feat(vaults): removed py-evm

sergeyWh1te committed:
feat(vaults): fix start point for calculation

sergeyWh1te committed:
feat(vaults): fix leafIndexToData

TheDZhon reviewed (commented):
:eyes:

sergeyWh1te committed:
feat(vaults): fix leafIndexToData

sergeyWh1te committed:
feat(vaults): fix lint

sergeyWh1te committed:
feat(vaults): fix lint

sergeyWh1te committed:
feat(vaults): fix lint

sergeyWh1te committed:
fix(vaults): lint moved StakingVaults to services folder

chasingrainbows committed:
feat(vaults): fix mypy

chasingrainbows committed:
feat(vaults): reformat test

tamtamchik committed:
chore(staking_vault): simplify pending deposits handling

sergeyWh1te committed:
feat(vaults): compact abies

tamtamchik committed:
test(staking_vault): calculate pending deposits happy path

sergeyWh1te committed:
Merge branch 'develop' into feat/vaults-pectra

# Conflicts:
#	src/modules/accounting/accounting.py

sergeyWh1te committed:
fix(vaults): ipfs test from latest block

sergeyWh1te committed:
fix(vault_fees): ignore previous report fee if vault reconnected between reports

If a vault was disconnected and later reconnected between two reports,
we no longer use the fee data from the previous report. The reason is that
the owner has already paid the outstanding liabilities, and the vault's
lifecycle effectively restarts from the reconnection point.

Closes: https://github.com/lidofinance/lido-otter/issues/733

sergeyWh1te committed:
fix(vault_fees): ignore previous report fee if vault reconnected between reports

If a vault was disconnected and later reconnected between two reports,
we no longer use the fee data from the previous report. The reason is that
the owner has already paid the outstanding liabilities, and the vault's
lifecycle effectively restarts from the reconnection point.

Closes: https://github.com/lidofinance/lido-oracle/issues/733

sergeyWh1te committed:
test(vault_fees): Cover by tests service_staking_vaults

sergeyWh1te committed:
fix(vaults): add check that `liability_shares` after VaultConnectedEvent must be zero

sergeyWh1te committed:
fix(vaults): slashingReserve in merkleTree must be greater 0. int256 -> uint256

sergeyWh1te committed:
fix(vaults): Prevent double-counting of vault events

sergeyWh1te committed:
tests(vaults): Test for getting events from prev_report_block_number + 1

sergeyWh1te committed:
tests(vaults): formatter

sergeyWh1te committed:
fix(vaults): review marks

sergeyWh1te committed:
fix(vaults): test formatter

chasingrainbows committed:
feat(orc-452): add car converter

chasingrainbows committed:
feat(orc-452): refactor to class

chasingrainbows committed:
feat(orc-452): add storacha provider

chasingrainbows committed:
feat(orc-452): remove star imports

chasingrainbows committed:
feat(orc-452): pylint - ignore protobuf files

chasingrainbows committed:
feat(orc-452): fix linter warnings

chasingrainbows committed:
feat(orc-452): allign mypy with protobuf

chasingrainbows committed:
feat(orc-452): add except on http errors

chasingrainbows committed:
feat(orc-452): add ipfs check

chasingrainbows committed:
feat(orc-452): add multiformats to dependencies

chasingrainbows committed:
feat(orc-452): remove PublicIPFS as not relevant

sergeyWh1te committed:
fix(vaults): Prevent negative apr

https://github.com/lidofinance/lido-oracle/issues/748

chasingrainbows committed:
feat(orc-452): replace ipfs multi provider with web3 IPFS module

chasingrainbows committed:
feat(orc-452): add get_frame_number_by_slot method to ConsensusModule

chasingrainbows committed:
feat(orc-452): add frame based ipfs providers rotation

chasingrainbows committed:
feat(orc-452): add ipfs rotation to vaults

chasingrainbows committed:
feat(orc-452): pass PINATA_JWT to CI

chasingrainbows committed:
feat(orc-452): add ipfs rotation to csm

sergeyWh1te committed:
fix(vaults): fix Invalid Core APR
https://github.com/lidofinance/lido-oracle/issues/753

chasingrainbows committed:
feat(orc-452): Preserve fallback state within the same frame if occurred

chasingrainbows committed:
feat(orc-452): refactor frame type

chasingrainbows committed:
feat(orc-452): add lru cache to get_frame_number_by_slot

chasingrainbows committed:
feat(orc-452): fix staking_vault/test_ifps_report.py

chasingrainbows committed:
feat(orc-452): add tests for web3py/extensions/ipfs.py

chasingrainbows committed:
feat(orc-452): Add new .env vars to readme

chasingrainbows committed:
feat(orc-452): add storacha envs to .env.example

chasingrainbows committed:
feat(orc-452): replace py-multiformats-cid with multiformats

chasingrainbows committed:
feat(orc-452): improve ipfs checker

chasingrainbows committed:
feat(orc-452): add tests for car converter

chasingrainbows committed:
feat(orc-452): add get_frame_number_by_slot test

chasingrainbows committed:
feat(orc-452): refact ipfs tests

chasingrainbows committed:
feat(orc-452): add storacha tests

chasingrainbows committed:
feat(orc-452): add CarFile dataclass

sergeyWh1te committed:
fix(vaults): fix invalid total value
https://github.com/lidofinance/lido-oracle/issues/754

sergeyWh1te committed:
fix(vaults): fix prevent negative lido fees
https://github.com/lidofinance/lido-oracle/issues/748

sergeyWh1te committed:
fix(vaults): lint black

sergeyWh1te committed:
fix(vaults): fix mypy warnings

chasingrainbows committed:
feat(orc-452): improve error readability

chasingrainbows committed:
feat(orc-452): Merge branch 'feat/vaults-pectra-audit-fixes' of github.com:lidofinance/lido-oracle into feature/orc-452-new-ipfs-providers-to-vault-oracles

chasingrainbows committed:
feat(orc-452): add retries to ipfs checks

chasingrainbows committed:
feat(orc-452): add cid validation on fetch and upload

chasingrainbows committed:
feat(orc-452): add pinata provider unit tests

sergeyWh1te committed:
fix(vaults): fix wrong LazyOracle abi.

sergeyWh1te committed:
fix(vaults): made stable sorting events if they come on the same block number

chasingrainbows committed:
feat(orc-452): add _normalize_cid method

sergeyWh1te committed:
fix(vaults): used pubkeys instead of deposit.signature in get_pending_deposits_by_vaults

sergeyWh1te committed:
fix(vaults): rename sort -> sort_events

sergeyWh1te committed:
fix(vaults): rename get_core_apr_ratio -> calculate_gross_core_apr

sergeyWh1te committed:
fix(vaults): use ascii instead of utf-8 for encoding  merkle_tree

sergeyWh1te committed:
fix(vaults): wrong block_hash and block_number for calculating start point for report

sergeyWh1te committed:
fix(vaults): rename vault_connected_events_set -> connected_vaults_set

sergeyWh1te committed:
fix(vaults): get_vaults_fees use liability_shares from prev report

sergeyWh1te committed:
fix(vaults): lint test_staking_vaults.py

chasingrainbows committed:
feat(orc-452): add specs

sergeyWh1te committed:
fix(vaults): lint test_staking_vaults.py

tamtamchik committed:
Merge pull request #734 from lidofinance/feat/vaults-pectra-audit-fixes

fix(vault_fees): Audit findings

chasingrainbows committed:
Merge branch 'feat/vaults-pectra' of github.com:lidofinance/lido-oracle into feature/orc-452-new-ipfs-providers-to-vault-oracles

chasingrainbows committed:
feat(orc-452): env IPFS_VALIDATE_CID added

chasingrainbows committed:
feat(orc-452): shuffle ipfs providers on startup

folkyatina commented on src/modules/accounting/accounting.py:
```diff
@@ -347,13 +371,108 @@ def _calculate_rebase_report(self, blockstamp: ReferenceBlockStamp) -> RebaseRep
+        core_apr_ratio = calculate_gross_core_apr(
+            pre_total_shares=simulation.pre_total_shares,
+            pre_total_ether=simulation.pre_total_pooled_ether,
+            post_total_shares=simulation.post_total_shares - simulation.shares_to_mint_as_fees,
```
we need to use `postInternalShares` and `postInternalEther` here to calculate gross apr correctly

sergeyWh1te committed:
Merge pull request #745 from lidofinance/feature/orc-452-new-ipfs-providers-to-vault-oracles

feat(orc-452): add storacha provider

sergeyWh1te committed:
fix(vaults): Removed get_all_validators unused const

F4ever committed:
Merge pull request #767 from lidofinance/develop

[EPIC] Triggerable Withdrawals (develop ⇒ master)

sergeyWh1te committed:
fix(vaults): Fixed calculation gross_apr

invalid-email-address committed:
chore(release): 6.1.0

F4ever committed:
Update pyproject.toml

F4ever committed:
Merge pull request #774 from lidofinance/pre-release-6.0.0

chore(release): 6.0.0

chasingrainbows committed:
feat(orc-455): Add Lido ipfs cluster integration

F4ever committed:
Merge pull request #778 from lidofinance/develop

Feat: bump curl version

invalid-email-address committed:
chore(release): 6.1.0

F4ever committed:
Update pyproject.toml

F4ever committed:
Merge pull request #780 from lidofinance/pre-release-6.0.0

chore(release): 6.0.0

chasingrainbows committed:
feat(orc-455): add tests for lido ipfs

chasingrainbows committed:
feat(orc-455): refactor ipfs checks

chasingrainbows committed:
feat(orc-455): implement ipfs "push to all" business logic

tamtamchik committed:
feat: redesign validator balance handling

tamtamchik committed:
chore: cleanup deps

tamtamchik committed:
chore: fix docker build

tamtamchik committed:
chore: fix linter

chasingrainbows committed:
feat(orc-455): bump curl

tamtamchik committed:
chore: simplify pending balances calculation

chasingrainbows committed:
feat(orc-455): remove frame num from ipfs.publish

tamtamchik committed:
chore: simplify code

tamtamchik committed:
chore: add comments and fix lint errors

tamtamchik committed:
ci: try to fix build

chasingrainbows committed:
feat(orc-455): reduce IPFS providers default timeout

invalid-email-address committed:
chore(release): 6.0.1

chasingrainbows committed:
feat(builds-1): feat(reproducible-builds): add rev to oz merkle tree and remove licenses on build

tamtamchik committed:
chore: update after review

tamtamchik committed:
chore: update after review

tamtamchik committed:
chore: update docs after review

tamtamchik committed:
Merge pull request #781 from lidofinance/feat/vaults-tv-calculation-update

feat: redesign validator balance handling

tamtamchik committed:
Merge pull request #776 from lidofinance/feature/orc-455-add-lido-ipfs-node-to-oracle

feat(orc-455): Add Lido ipfs cluster integration

chasingrainbows committed:
feat(builds-1): feat(reproducible-build): add version to Dockerfile to fix reproducible builds, according to https://github.com/moby/buildkit/issues/6217

F4ever committed:
Merge pull request #785 from lidofinance/feature/hotfix-for-reproducibble-builds-1

Feature/hotfix for reproducibble builds

F4ever committed:
Merge pull request #783 from lidofinance/pre-release-6.0.1

chore(release): 6.0.1

tamtamchik committed:
chore(vaults): fix pdg flow

tamtamchik committed:
chore(vaults): formating

tamtamchik committed:
chore(vaults): abidiff

tamtamchik committed:
feat(vaults): restore share_rate

tamtamchik committed:
feat(vaults): add max_liability_shares

tamtamchik committed:
feat(vaults): use new locator

tamtamchik committed:
chore: fixes

tamtamchik committed:
feat(vaults): merge master

tamtamchik committed:
Merge branch 'feat/fix-for-pdg-flow' into feat/fix-775

# Conflicts:
#	src/modules/accounting/types.py

tamtamchik committed:
feat(vaults): minify abis

tamtamchik committed:
chore: after review

tamtamchik committed:
feat(vaults): fix or skip tests

tamtamchik committed:
Merge pull request #787 from lidofinance/feat/fix-775

feat: fix share_rate and add max liability shares

tamtamchik committed:
feat: use lazy oracle to check validator stages

tamtamchik committed:
chore: please linter

tamtamchik committed:
feat: add integration tests for lazy oracle

tamtamchik committed:
Merge remote-tracking branch 'origin/fix/finalization-batches-build' into feat/audits-3

# Conflicts:
#	pyproject.toml

tamtamchik committed:
Merge branch 'feat/audits-3' into feat/validator-stages

tamtamchik committed:
chore: apply review comments

tamtamchik committed:
Merge branch 'develop' into feat/validator-stages

# Conflicts:
#	pyproject.toml

tamtamchik committed:
chore: apply review comments

tamtamchik committed:
chore: fixes after review

tamtamchik committed:
Merge pull request #790 from lidofinance/feat/validator-stages

feat: use lazy oracle to check validator stages

tamtamchik committed:
fix: h-02 – double counting

tamtamchik committed:
feat(vaults): fix test

hweawer committed:
Fix csm test

tamtamchik committed:
feat: bump ao contract version

tamtamchik committed:
Merge pull request #784 from lidofinance/feat/vaults-testnet-3

[VAULTS]: Audit fixes 2

tamtamchik commented on src/modules/accounting/accounting.py:
```diff
@@ -57,8 +77,8 @@ class Accounting(BaseModule, ConsensusModule):
-    COMPATIBLE_CONTRACT_VERSION = 2
-    COMPATIBLE_CONSENSUS_VERSION = 4
+    COMPATIBLE_CONTRACT_VERSION = 3
+    COMPATIBLE_CONSENSUS_VERSION = 3
```
Fixed in bb0a1911289e1077223a27b0ed0ecdb4523f037d

TheDZhon commented on src/modules/csm/csm.py:
why do we touch this file?

tamtamchik replied:
I see it was part of the IPFS update process, specifically, the `_get_last_report` function has been changed.

TheDZhon commented on src/modules/csm/distribution.py:
why we change this file?

tamtamchik replied:
This is just formatting, would leave as it is now

TheDZhon commented on src/providers/ipfs/kubo.py:
shall we remove kubo though?

TheDZhon reviewed (approved):
LGTM, just a few non-blocking questions
